zcash_client_backend/src/keys.rs

```diff
@@ -536,19 +537,46 @@ impl UnifiedFullViewingKey {
         self.orchard.as_ref()
     }
 
-    /// Attempts to derive the Unified Address for the given diversifier index.
+    /// Attempts to derive the Unified Address for the given diversifier index and
+    /// receiver types.
     ///
     /// Returns `None` if the specified index does not produce a valid diversifier.
-    // TODO: Allow filtering down by receiver types?
-    pub fn address(&self, j: DiversifierIndex) -> Option<UnifiedAddress> {
-        let sapling = if let Some(extfvk) = self.sapling.as_ref() {
+    pub fn address(
+        &self,
+        j: DiversifierIndex,
+        receiver_types: &[Typecode],
+    ) -> Option<UnifiedAddress> {
```

Comment:
We need to confirm that `receiver_types` only contains `Typecode`s for which we can generate receivers. If an unsupported `Typecode` is requested, we should fail:
```suggestion
    ) -> Option<UnifiedAddress> {
        // If the user requests an unsupported receiver type, we can't produce an address.
        if receiver_types.iter()
            .any(|typecode| match typecode {
                #[cfg(feature = "orchard")]
                Typecode::Orchard => false,
                Typecode::Sapling => false,
                #[cfg(feature = "transparent-inputs")]
                Typecode::P2pkh => false,
                _ => true,
            })
        {
            return None;
        }

```
Previously this wasn't a problem because we encoded a default preference, but now that the caller has to be explicit, we need to respect that explicitness.

We might also want to make this `Option<&[Typecode]>` so if the caller specifies `None` we do the old default behaviour.

Reply:
This makes sense. I think though that instead of returning `Option<UnifiedAddress>` we should at this point probably then switch to a `Result` with a bespoke error enum, since generation can fail in multiple ways.